sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java

```diff
@@ -109,56 +109,61 @@ public PartitionControllerImpl(
-                    }
 
-                    return Mono.empty();
-                })
-                .doOnSuccess(aVoid -> {
-                    logger.info("Partition {}: successfully removed lease.", lease.getLeaseToken());
-                });
+                return this.leaseManager.release(lease);
+            })
+            .onErrorResume(e -> {
+                if (e instanceof LeaseLostException) {
+                    logger.warn("Partition {}: lease already removed.", lease.getLeaseToken());
+                } else {
+                    logger.warn("Partition {}: failed to remove lease.", lease.getLeaseToken(), e);
+                }
+
+                return Mono.empty();
+            })
+            .doOnSuccess(aVoid -> {
+                logger.info("Partition {}: successfully removed lease.", lease.getLeaseToken());
+            });
     }
 
     private WorkerTask processPartition(PartitionSupervisor partitionSupervisor, Lease lease) {
         CancellationToken cancellationToken = this.shutdownCts.getToken();
 
-        WorkerTask partitionSupervisorTask = new WorkerTask(lease, () -> {
-            partitionSupervisor.run(cancellationToken)
-                .onErrorResume(throwable -> {
-                    if (throwable instanceof PartitionSplitException) {
-                        PartitionSplitException ex = (PartitionSplitException) throwable;
-                        return this.handleSplit(lease, ex.getLastContinuation());
-                    } else if (throwable instanceof TaskCancelledException) {
-                        logger.debug("Partition {}: processing canceled.", lease.getLeaseToken());
-                    } else {
-                        logger.warn("Partition {}: processing failed.", lease.getLeaseToken(), throwable);
-                    }
-
-                    return Mono.empty();
-                })
-                .then(this.removeLease(lease)).subscribe();
-        });
+        WorkerTask partitionSupervisorTask =
+            new WorkerTask(
+                lease,
+                shutdownCts,
+                getWorkerJob(partitionSupervisor, lease, cancellationToken));
 
         this.scheduler.schedule(partitionSupervisorTask);
 
         return partitionSupervisorTask;
     }
 
+    private Mono<Void> getWorkerJob(PartitionSupervisor partitionSupervisor, Lease lease, CancellationToken cancellationToken) {
+        return partitionSupervisor.run(cancellationToken)
+            .onErrorResume(throwable -> {
+                if (throwable instanceof PartitionSplitException) {
+                    PartitionSplitException ex = (PartitionSplitException) throwable;
+                    return this.handleSplit(lease, ex.getLastContinuation());
+                } else if (throwable instanceof TaskCancelledException) {
+                    logger.debug("Partition {}: processing canceled.", lease.getLeaseToken());
```

Comment:
info? Given how verbose we are in CFP on info level and how much it was needed to have the logs to understand what is going on?

Reply:
never mind - not needed here...